### PR TITLE
Convert gradle build scripts to kts

### PR DESCRIPTION
**Pull request by Goooler**
## Description

Follow up #1428, convert Gradle build scripts to KTS to be type-safe.

https://docs.gradle.org/current/userguide/kotlin_dsl.html

## Checklist

- [x] PR description added
- [ ] tests are added
- [ ] `CHANGELOG.md` is updated

In case of adding a new rule:
- [ ] `README.md` is updated
- [ ] Rule has been applied on Ktlint itself and violations are fixed

### CODE REVIEW

File: ktlint-ruleset-template/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    testImplementation(projects.ktlintTest)
+    testImplementation(libs.junit5)
+    testImplementation(libs.assertj)
+}
+
+tasks.test {
+    useJUnitPlatform()
+}
+
+tasks.register<JavaExec>("ktlint") {
```
Comment (Goooler) [inline ktlint-ruleset-template/build.gradle.kts:43]: The `ktlint` task's configuration in `ktlint-ruleset-template` is not same with `ktlint` task in root `build.gradle`, I'm not sure which one is right.

https://github.com/pinterest/ktlint/blob/c42647182196339f08ff14cea9dd0321d69014fc/build.gradle#L30-L39

File: ktlint-ruleset-template/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    dependsOn(tasks.classes)
+    group = LifecycleBasePlugin.VERIFICATION_GROUP
+    mainClass.set("com.pinterest.ktlint.Main")
+    // adding compiled classes to the classpath so that ktlint would validate project"s sources
+    // using its own ruleset (in other words to dogfood)
+    classpath = ktlint + sourceSets.main.get().output
+    args("--debug", "src/**/*.kt")
+}.let {
+    tasks.check.get().dependsOn(it)
+}
```
Comment (Goooler) [inline ktlint-ruleset-template/build.gradle.kts:53]: The `ktlint` task's configuration in `ktlint-ruleset-template` is not same with `ktlint` task in root `build.gradle`, I'm not sure which one is right.

https://github.com/pinterest/ktlint/blob/c42647182196339f08ff14cea9dd0321d69014fc/build.gradle#L30-L39

Comment (Goooler) [reply ktlint-ruleset-template/build.gradle.kts:53]: The main diffs are `dependsOn`.

File: ktlint-reporter-sarif/build.gradle [hunk 3]
```diff
@@ -1,18 +0,0 @@
...
-
-  testImplementation libs.junit5
-  testImplementation libs.assertj
-}
-
-jar {
-    manifest {
-        attributes 'Implementation-Version': project.property('VERSION_NAME')
-    }
-}
```
Comment (Goooler) [inline ktlint-reporter-sarif/build.gradle:18]: I'm removing the `jar` block as it is showing in `ktlint` module, right?

https://github.com/Goooler/ktlint/blob/2f7587351e8e57d30148683ab4fc0ea9d3c12a49/ktlint/build.gradle.kts#L11-L16

File: build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    classpath = ktlint
+    mainClass.set("com.pinterest.ktlint.Main")
+    // Experimental rules run by default run on the ktlint code base itself. Experimental rules should not be released if
+    // we are not pleased ourselves with the results on the ktlint code base.
+    // Sources in "ktlint/src/test/resources" are excluded as those source contain lint errors that have to be detected by
+    // unit tests and should not be reported/fixed.
+    args(
+        "**/src/**/*.kt",
+        "**.kts",
+        "!**/build/**",
```
Comment (Goooler) [inline build.gradle.kts:39]: Newly added rules, `ktlint` task doesn't work on kts files before.

File: buildSrc/src/main/kotlin/ktlint-kotlin-common.gradle.kts [hunk 5]
```diff
@@ -19,11 +19,10 @@ kotlin {
     }
 }
 
-tasks
-    .withType<KotlinCompile>()
+tasks.withType<KotlinCompile>()
     .configureEach {
         kotlinOptions {
-            apiVersion = "1.4"
+            apiVersion = "1.6"
```
Comment (Goooler) [inline buildSrc/src/main/kotlin/ktlint-kotlin-common.gradle.kts:25]:
```
w: API version 1.4 is deprecated and its support will be removed in a future version of Kotlin
```

File: ktlint-ruleset-template/build.gradle.kts (hunk 2 above)
Comment (paul-dingemans) [reply ktlint-ruleset-template/build.gradle.kts:53]: The root `build.gradle` is definitively the correct one. When you run it, you can see that it parses about 220+ files. I expect that the version in `ktlint-ruleset-template` is meant as example for developers who want to write their own custom ruleset.

File: ktlint-reporter-sarif/build.gradle (hunk 3 above)
Comment (paul-dingemans) [reply ktlint-reporter-sarif/build.gradle:18]: @romtsn @shashachu I can not judge this.

File: ktlint-ruleset-template/build.gradle.kts (hunk 2 above)
Comment (Goooler) [reply ktlint-ruleset-template/build.gradle.kts:53]: Ok.

File: buildSrc/src/main/kotlin/ktlint-kotlin-common.gradle.kts (hunk 5 above)
Comment (shashachu) [reply buildSrc/src/main/kotlin/ktlint-kotlin-common.gradle.kts]: I don't completely know the consequences of us increasing this apiVersion. Will some plugins etc become incompatible? @Tapchicoma ?

Comment (paul-dingemans) [reply buildSrc/src/main/kotlin/ktlint-kotlin-common.gradle.kts]: @Goooler In [commit](https://github.com/pinterest/ktlint/pull/1449/commits/311b63d6ba7fcc4d3b5bed36c27f4a0089a3b1cd) you have removed the `apiVersion` entirely. What is the reasoning behind this?

Comment (Goooler) [reply buildSrc/src/main/kotlin/ktlint-kotlin-common.gradle.kts]: As the apiVersion is sync with Kotlin version by default.

**Comment by paul-dingemans** [comment]
Ok, lets give this change a go and hope for the best. 🤞

**Comment by paul-dingemans** [comment]
@Goooler Tnx for your patience on this issue. Your contributions are welcome. Unfortunately, review and merging sometimes take a bit longer as I have little knowledge about these kind of changes.

**Pull request merged**